File: SoftAndSearch.py
Selection Sort
def selection_sort(data):
    """Сортировка выбором — ищем минимальный элемент и ставим его в начало."""
    n = len(data)
    for i in range(n - 1):
        min_pos = i
        for j in range(i + 1, n):
            if data[j] < data[min_pos]:
                min_pos = j
        if min_pos != i:
            data[i], data[min_pos] = data[min_pos], data[i]
    return data
🔹 Bubble Sort
def bubble_sort(data):
    """Пузырьковая сортировка — многократно обходим массив, меняя местами соседние элементы."""
    length = len(data)
    for i in range(length):
        swapped = False
        for j in range(0, length - i - 1):
            if data[j] > data[j + 1]:
                data[j], data[j + 1] = data[j + 1], data[j]
                swapped = True
        if not swapped:
            break
    return data
🔹 Insertion Sort
def insertion_sort(data):
    """Сортировка вставками — каждый элемент вставляется в нужное место в отсортированной части."""
    for i in range(1, len(data)):
        current = data[i]
        j = i - 1
        while j >= 0 and data[j] > current:
            data[j + 1] = data[j]
            j -= 1
        data[j + 1] = current
    return data
🔹 Merge Sort
def merge_sort(data):
    """Сортировка слиянием — делим массив пополам и сливаем отсортированные части."""
    if len(data) <= 1:
        return data

    middle = len(data) // 2
    left = merge_sort(data[:middle])
    right = merge_sort(data[middle:])
    return merge(left, right)

def merge(a, b):
    """Слияние двух отсортированных списков."""
    result = []
    i = j = 0
    while i < len(a) and j < len(b):
        if a[i] <= b[j]:
            result.append(a[i])
            i += 1
        else:
            result.append(b[j])
            j += 1
    result.extend(a[i:])
    result.extend(b[j:])
    return result
🔹 Shell Sort
def shell_sort(data):
    """Сортировка Шелла — выполняем сортировку вставками с уменьшающимся шагом."""
    n = len(data)
    gap = n // 2
    while gap > 0:
        for i in range(gap, n):
            temp = data[i]
            j = i
            while j >= gap and data[j - gap] > temp:
                data[j] = data[j - gap]
                j -= gap
            data[j] = temp
        gap //= 2
    return data
🔹 Quick Sort
def quick_sort(data):
    """Быстрая сортировка — разделяем элементы относительно опорного."""
    if len(data) < 2:
        return data

    pivot = data[len(data) // 2]
    less = [x for x in data if x < pivot]
    equal = [x for x in data if x == pivot]
    greater = [x for x in data if x > pivot]

    return quick_sort(less) + equal + quick_sort(greater)
🔹 Heap Sort
def heap_sort(data):
    """Пирамидальная сортировка — строим кучу и извлекаем максимумы."""
    n = len(data)

    for i in range(n // 2 - 1, -1, -1):
        _heapify(data, n, i)

    for end in range(n - 1, 0, -1):
        data[0], data[end] = data[end], data[0]
        _heapify(data, end, 0)

    return data

def _heapify(arr, size, root):
    largest = root
    left = 2 * root + 1
    right = 2 * root + 2

    if left < size and arr[left] > arr[largest]:
        largest = left
    if right < size and arr[right] > arr[largest]:
        largest = right

    if largest != root:
        arr[root], arr[largest] = arr[largest], arr[root]
        _heapify(arr, size, largest)
🔹 Linear Search
def linear_search(data, key):
    """Последовательный поиск — проверяем элементы один за другим."""
    for i, value in enumerate(data):
        if value == key:
            return i
    return -1
🔹 Binary Search
def binary_search(data, key):
    """Бинарный поиск — делим область поиска пополам."""
    left, right = 0, len(data) - 1
    while left <= right:
        mid = (left + right) // 2
        if data[mid] == key:
            return mid
        elif data[mid] < key:
            left = mid + 1
        else:
            right = mid - 1
    return -1
🔹 Interpolation Search
def interpolation_search(data, key):
    """Интерполяционный поиск — улучшенный бинарный, позиция вычисляется по формуле."""
    low, high = 0, len(data) - 1

    while low <= high and key >= data[low] and key <= data[high]:
        if data[low] == data[high]:
            return low if data[low] == key else -1

        pos = low + ((key - data[low]) * (high - low)) // (data[high] - data[low])

        if data[pos] == key:
            return pos
        elif data[pos] < key:
            low = pos + 1
        else:
            high = pos - 1
    return -1
🔹 Fibonacci Search
def fibonacci_search(data, key):
    """Поиск Фибоначчи — делит массив по числам Фибоначчи."""
    n = len(data)
    fib2, fib1 = 0, 1
    fib = fib1 + fib2

    while fib < n:
        fib2, fib1 = fib1, fib
        fib = fib1 + fib2

    offset = -1

    while fib > 1:
        i = min(offset + fib2, n - 1)
        if data[i] < key:
            fib, fib1, fib2 = fib1, fib2, fib - fib1
            offset = i
        elif data[i] > key:
            fib, fib1, fib2 = fib2, fib1 - fib2, fib - fib1
        else:
            return i

    if fib1 == 1 and offset + 1 < n and data[offset + 1] == key:
        return offset + 1
    return -1

File: SoftAndSearch.java
SelectionSort (Сортировка выбором)
public class SelectionSort {
    public static void sort(int[] array) {
        int n = array.length;

        for (int i = 0; i < n - 1; i++) {
            int minPos = i;

            for (int j = i + 1; j < n; j++) {
                if (array[j] < array[minPos]) {
                    minPos = j;
                }
            }

            if (minPos != i) {
                int temp = array[i];
                array[i] = array[minPos];
                array[minPos] = temp;
            }
        }
    }
}
🔹 BubbleSort (Пузырьковая)
public class BubbleSort {
    public static void sort(int[] arr) {
        int len = arr.length;
        boolean swapped;

        for (int i = 0; i < len - 1; i++) {
            swapped = false;

            for (int j = 0; j < len - i - 1; j++) {
                if (arr[j] > arr[j + 1]) {
                    int tmp = arr[j];
                    arr[j] = arr[j + 1];
                    arr[j + 1] = tmp;
                    swapped = true;
                }
            }

            if (!swapped) break; // Если обменов не было, массив уже отсортирован
        }
    }
}
🔹 InsertionSort (Вставками)
public class InsertionSort {
    public static void sort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int current = arr[i];
            int j = i - 1;

            while (j >= 0 && arr[j] > current) {
                arr[j + 1] = arr[j];
                j--;
            }

            arr[j + 1] = current;
        }
    }
}
🔹 MergeSort (Слиянием)
public class MergeSort {
    public static void sort(int[] arr, int left, int right) {
        if (left >= right) return;

        int mid = (left + right) / 2;
        sort(arr, left, mid);
        sort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }

    private static void merge(int[] arr, int left, int mid, int right) {
        int[] leftArr = new int[mid - left + 1];
        int[] rightArr = new int[right - mid];

        for (int i = 0; i < leftArr.length; i++) leftArr[i] = arr[left + i];
        for (int j = 0; j < rightArr.length; j++) rightArr[j] = arr[mid + 1 + j];

        int i = 0, j = 0, k = left;

        while (i < leftArr.length && j < rightArr.length) {
            arr[k++] = (leftArr[i] <= rightArr[j]) ? leftArr[i++] : rightArr[j++];
        }

        while (i < leftArr.length) arr[k++] = leftArr[i++];
        while (j < rightArr.length) arr[k++] = rightArr[j++];
    }
}
🔹 ShellSort
public class ShellSort {
    public static void sort(int[] arr) {
        int n = arr.length;

        for (int gap = n / 2; gap > 0; gap /= 2) {
            for (int i = gap; i < n; i++) {
                int temp = arr[i];
                int j = i;

                while (j >= gap && arr[j - gap] > temp) {
                    arr[j] = arr[j - gap];
                    j -= gap;
                }

                arr[j] = temp;
            }
        }
    }
}
🔹 QuickSort (Быстрая)
public class QuickSort {
    public static void sort(int[] arr, int low, int high) {
        if (low >= high) return;

        int pivotIndex = partition(arr, low, high);
        sort(arr, low, pivotIndex - 1);
        sort(arr, pivotIndex + 1, high);
    }

    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int idx = low - 1;

        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                idx++;
                int tmp = arr[idx];
                arr[idx] = arr[j];
                arr[j] = tmp;
            }
        }

        int tmp = arr[idx + 1];
        arr[idx + 1] = arr[high];
        arr[high] = tmp;

        return idx + 1;
    }
}
🔹 HeapSort (Пирамидальная)
public class HeapSort {
    public static void sort(int[] arr) {
        int n = arr.length;

        for (int i = n / 2 - 1; i >= 0; i--) heapify(arr, n, i);
        for (int i = n - 1; i > 0; i--) {
            int tmp = arr[0];
            arr[0] = arr[i];
            arr[i] = tmp;

            heapify(arr, i, 0);
        }
    }

    private static void heapify(int[] arr, int size, int root) {
        int largest = root;
        int left = 2 * root + 1;
        int right = 2 * root + 2;

        if (left < size && arr[left] > arr[largest]) largest = left;
        if (right < size && arr[right] > arr[largest]) largest = right;

        if (largest != root) {
            int tmp = arr[root];
            arr[root] = arr[largest];
            arr[largest] = tmp;

            heapify(arr, size, largest);
        }
    }
}
🔹 LinearSearch
public class LinearSearch {
    public static int find(int[] arr, int value) {
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] == value) return i;
        }
        return -1;
    }
}
🔹 BinarySearch
public class BinarySearch {
    public static int find(int[] arr, int value) {
        int left = 0;
        int right = arr.length - 1;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (arr[mid] == value) return mid;
            if (arr[mid] < value) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }
}
🔹 InterpolationSearch
public class InterpolationSearch {
    public static int find(int[] arr, int key) {
        int low = 0, high = arr.length - 1;

        while (low <= high && key >= arr[low] && key <= arr[high]) {
            if (arr[low] == arr[high]) {
                return arr[low] == key ? low : -1;
            }

            int pos = low + ((key - arr[low]) * (high - low)) / (arr[high] - arr[low]);

            if (arr[pos] == key) return pos;
            if (arr[pos] < key) low = pos + 1;
            else high = pos - 1;
        }
        return -1;
    }
}
🔹 FibonacciSearch
public class FibonacciSearch {
    public static int find(int[] arr, int key) {
        int n = arr.length;
        int fib2 = 0, fib1 = 1, fib = fib1 + fib2;

        while (fib < n) {
            fib2 = fib1;
            fib1 = fib;
            fib = fib1 + fib2;
        }

        int offset = -1;

        while (fib > 1) {
            int i = Math.min(offset + fib2, n - 1);

            if (arr[i] < key) {
                fib = fib1;
                fib1 = fib2;
                fib2 = fib - fib1;
                offset = i;
            } else if (arr[i] > key) {
                fib = fib2;
                fib1 = fib1 - fib2;
                fib2 = fib - fib1;
            } else {
                return i;
            }
        }

        if (fib1 == 1 && offset + 1 < n && arr[offset + 1] == key) return offset + 1;
        return -1;
    }
}
